### ISSUE TITLE
Stable Spartanization

### ISSUE THREAD

**Issue by OriRoth** [SpartanRefactoring/Spartanizer#48]
Starting building stable spartanization. This issue contains all relevant material about the subject:

Known issues need to be solved:
  * ternary type mismatch (see http://bugs.java.com/view_bug.do?bug_id=6721089)
  * Disabled assignments consolidation for NumberLiterals values
  * Further inspection of changed Is#blockEssential is needed

Issues answered:
  * inlining into ClassInstanceCreation node (see DeclarationInitializerReturnExpression) - the new method looks for illegal parents of variable use (such as ClassInstanceCreation), which is sufficient. Maybe needed in other places - the correction is mainly in LocalInlineWithValue#canSafelyInlineInto
  * inline array initialization - not allowed (see DeclarationInitializerStatementTerminatingScope)
  * extraction of common suffix with undefined variables - changed IfBarFooElseBazFoo#make = added a visit of the common suffix to check for variables definitions in prefix, using Collect.definitionsOf (that I made for this purpose)
  * safer BlockSimplify wring - checks for duplicate variables declarations, see BlockSimplify#safeBlockSimplification
  * two assignments same value different types (such as `int x;double y;x=0;y=0;`) - disabled the option for NumberLiteral assignments (cannot be resolved without binding resolution)
  * No if to ternary conversion that creates a MethodInvocation (maybe could be done better using binding)
  * Dropping of redundant brackets - recursive inspection for if-no-else for blockEssential wring in Is#blockEssential

Also new tests for stabilization, see TrimmerTest

Fixes in original code:
  * Changed Collect.forAllOccurencesExcludingDefinitions to Collect.usesOf in Wring#inlineIntoSingleton: original intention unclear
  * Is#blockEssential - replaced else statement extraction with recursive implementation Funcs#recursiveElze

Notice:
  * Disabled inlining of final variables
  * Ternary pushdown disabled for MethodInvocation

**Comment by OriRoth** [comment]
There is a problem with turning if statements into ternary expression. This is a very important wring, but in some cases it fails to type mismatch (http://stackoverflow.com/questions/19305095/type-mismatch-cannot-convert-from-object-to-k-when-using-conditional-operator for example). Should I resolve the type of the expressions, or just ignore the problem?
@yossigil

**Comment by OriRoth** [comment]
What about efficiency issues? It might be important in a mathematical library:
For example, should
```
a = new double[N][N];
for (int j = 0; j < N; j++) {
  final double[] aJ = a[j];
  for (int i = 0; i < N; i++) {
    aJ[i] = coeff[i * N + j];
  }
}
```
be changed to
```
a = new double[N][N];
for (int j = 0; j < N; j++) {
  for (int i = 0; i < N; i++) {
    a[j][i] = coeff[i * N + j];
  }
}
```
?

**Comment by OriRoth** [comment]
**Precision**:
In some cases, innocent spartanization such as InfixSortAddition (changes order of plus operations) can break the tests of the math library, that checks precision that naturally affected by order of operations. For instance, InfixSortAddition will suggest the transformation
`final double coeff1    = s * (6 - 9 * theta + fourTheta2);`
==>
`final double coeff1    = s * (fourTheta2 + 6 - 9 * theta);`
in org.apache.commons.math3.ode.nonstiff.GillStepInterpolator line 116
that breaks the test GillFieldStepInterpolatorTest.
Notice that most of the breaking in tests in common math project derive from wrings for InfixExpressions (mostly reordering of arithmetical operations), which probably defect precision.

@yossigil should we ignore the precision matter in general and just disable this spartanization option for common math project?

**Comment by yossigil** [comment]
Matteo, I think we should look for another Maven project.

**Comment by yossigil** [comment]
@OriRoth Not sure. We may need to search for another project which is better covered in tests?

**Comment by yossigil** [comment]
I think @M4tt30rru  already started making  this search

**Comment by OriRoth** [comment]
**Midway Conclusion**
*Comments written here nullify the above*
*Relevant for branch more-stable only*
In order to make the Spartanization tool safe but still usable, some security measures were removed. Now I will describe the problematic issues:

1. AssignmentAndAssignment not type safe: `x = 0; y = 0;` ==> `x = y = 0;` for example can cause compilation problems: `int x; double y; x = 0; y = 0;` =\=> `x = y = 0;`. This problem can easily be solved using *REFLECTIONS*. The non-reflection solution I used was to disable the wring when the assignment value is NumberLiteral (because one cannot know what the compiler would consider it without reflections).
2. If to ternary with method invocation - as said above, turning an IfStatement into a ConditionalExpression can cause compilation problems when using generics (known Java bug, see link above). For usability issues, this feature is no longer disabled. The problem may be solved using reflections.
3. Is#BlockEssential: I did not reverted to the original implementation, because I added a few test cases that fail it. There should be a review of the subject.

Furthermore: all tests pass after deleting irrelevant spartanizations and tests (see 74992a14a6301a1ff04fb0dca0fccf5c549ad577). This version is more stable than before, but can still cause compilation problems along with logical errors.
